### PR TITLE
Bear.py: Add attributes for setup.py

### PR DESCRIPTION
**Pull request by Adrianzatreanu**
This helps maintaining all the attributes
in each Bear, to be able to generate a
setup.py accordingly for each bear.

### CODE REVIEW

File: coalib/bears/Bear.py [hunk 1]
```diff
@@ -84,6 +110,14 @@ class Bear(Printer, LogPrinter):
 
     LANGUAGES = ()
     REQUIREMENTS = ()
+    AUTHOR = ""
+    AUTHOR_EMAIL = ""
+    MAINTAINER = ()
+    MAINTAINER_EMAIL = ()
+    PLATFORMS = 'any'
+    PACKAGE_DATA = {}
```
Comment (sils1297) [inline coalib/bears/Bear.py]: why package data?

File: coalib/bears/Bear.py [hunk 2]
```diff
@@ -84,6 +110,14 @@ class Bear(Printer, LogPrinter):
...
+    LICENSE = ""
+    ENTRY_POINTS = {"coalabears": ["coala_official_bears = bears"]}
```
Comment (sils1297) [inline coalib/bears/Bear.py]: nope, this shouldn't be defined by the user

File: coalib/bears/Bear.py [hunk 3]
```diff
@@ -84,6 +106,12 @@ class Bear(Printer, LogPrinter):
 
     LANGUAGES = ()
     REQUIREMENTS = ()
+    AUTHOR = ""
+    AUTHOR_EMAIL = ""
+    MAINTAINER = ()
+    MAINTAINER_EMAIL = ()
```
Comment (sils1297) [inline coalib/bears/Bear.py]: why's this a tuple and author a string?

Comment (Adrianzatreanu) [inline coalib/bears/Bear.py]: because there can be only one author, but multiple maintainers

File: coalib/bears/Bear.py [hunk 4]
```diff
@@ -80,7 +80,7 @@ class Bear(Printer, LogPrinter):
     ...     AUTHOR_EMAIL = "jon_snow@gmail.com"
     ...     MAINTAINER = ("Catelyn Stark",)
     ...     MAINTAINER_EMAIL = ("catelyn_stark@gmail.com",)
-    ...     PACKAGE_DATA = {'bears.java': ['checkstyle.jar',
+    ...     INCLUDE_LOCAL_FILES = {'bears.java': ['checkstyle.jar',
```
Comment (sils1297) [inline coalib/bears/Bear.py]: wrong commit

File: coalib/bears/Bear.py [hunk 5]
```diff
@@ -95,6 +95,24 @@ class Bear(Printer, LogPrinter):
...
+    >>> CAN_FIX = ('Syntax', ...)
+
+    >>> class SomeBear(Bear):
+    ...     CAN_DETECT = ('Syntax', 'Formatting')
+    ...     CAN_FIX = ('Syntax',)
+
+    >>> SomeBear.CAN_DETECT
+    ('Syntax', 'Formatting')
+    >>> SomeBear.CAN_FIX
+    ('Syntax',)
```
Comment (sils1297) [inline coalib/bears/Bear.py]: I'd remove 108-115

File: coalib/bears/Bear.py [hunk 6]
```diff
@@ -74,6 +74,28 @@ class Bear(Printer, LogPrinter):
...
 
+    To specify additional attributes to your bear, use the following:
+    >>> class SomeBear(Bear):
+    ...     AUTHOR = "Jon Snow"
+    ...     AUTHOR_EMAIL = "jon_snow@gmail.com"
+    ...     MAINTAINER = ("Catelyn Stark",)
+    ...     MAINTAINER_EMAIL = ("catelyn_stark@gmail.com",)
+    ...     PACKAGE_DATA = {'bears.java': ['checkstyle.jar',
+    ...                                 'google_checks.xml']}
+    ...     LICENSE = "AGPL-3.0"
```
Comment (sils1297) [inline coalib/bears/Bear.py]: remove the tests for invoking the argument, they're not needed and not helpful

File: coalib/bears/Bear.py [hunk 7]
```diff
@@ -74,6 +74,17 @@ class Bear(Printer, LogPrinter):
     ...     REQUIREMENTS = PythonRequirement.multiple(
     ...         ('colorama', '0.1'), 'coala_decorators')
 
+    To specify additional attributes to your bear, use the following:
+    
+    >>> class SomeBear(Bear):
+    ...     AUTHOR = "Jon Snow"
+    ...     AUTHOR_EMAIL = "jon_snow@gmail.com"
+    ...     MAINTAINER = ("Catelyn Stark",)
+    ...     MAINTAINER_EMAIL = ("catelyn_stark@gmail.com",)
```
Comment (sils1297) [inline coalib/bears/Bear.py]: maintainerS.

Also you want to have a property for this which always makes it a list, just like we do with LANGUAGES

File: coalib/bears/Bear.py [hunk 8]
```diff
@@ -74,6 +74,17 @@ class Bear(Printer, LogPrinter):
...
 
+    To specify additional attributes to your bear, use the following:
+
+    >>> class SomeBear(Bear):
+    ...     AUTHOR = "Jon Snow"
+    ...     AUTHOR_EMAIL = "jon_snow@gmail.com"
+    ...     MAINTAINERS = ("Catelyn Stark",)
+    ...     MAINTAINERS_EMAILS = ("catelyn_stark@gmail.com",)
+    ...     INCLUDE_LOCAL_FILES = {'bears.java': ['checkstyle.jar',
+    ...                                 'google_checks.xml']}
```
Comment (sils1297) [inline coalib/bears/Bear.py]: this needs documentation, why is it this syntax and not simply paths to files? Make it simpler please :)

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2338/commits/5fd1a38078f175e91b46f09f8e25222ffdcf343d if it passes

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2338/commits/5b65e9b2825f5ebe1222f31644f861a400f6563c

**Comment by sils1297** [comment]
third commit missing, correcting the supported languages property

**Comment by Adrianzatreanu** [comment]
sure, in a moment

**Comment by Adrianzatreanu** [comment]
@sils1297 seems theres an issue with gitmate, sometimes it throws me issues from commits that are not even there. this usually happens when rebasing, so it might be from the MERGE commit that is there but not shown, not 100% sure.

File: coalib/bears/Bear.py [hunk 9]
```diff
@@ -100,6 +130,30 @@ def supported_languages(cls):
...
+        return ((cls.MAINTAINERS_EMAILS,) if isinstance(
+            cls.MAINTAINERS, str) else cls.MAINTAINERS_EMAILS)
+
+    @classproperty
+    def include_local_files(cls):
+        """
+        :return: A tuple with local file paths to be included.
+        """
+        return ((cls.INCLUDE_LOCAL_FILES,) if isinstance(
+            cls.INCLUDE_LOCAL_FILES, str) else cls.INCLUDE_LOCAL_FILES)
```
Comment (sils1297) [inline coalib/bears/Bear.py]: oh yes for those you'll need tests as they're functions, which aren't evaluated on module load right?

File: coalib/bears/Bear.py [hunk 10]
```diff
@@ -74,6 +74,35 @@ class Bear(Printer, LogPrinter):
...
+    ...     INCLUDE_LOCAL_FILES = "checkstyle.jar"
+    >>> SomeBear.maintainers
+    ('Catelyn Stark',)
+    >>> SomeBear.maintainers_emails
+    ('catelyn_stark@gmail.com',)
+    >>> SomeBear.include_local_files
+    ('checkstyle.jar',)
+
+    If your bear needs to include local files, then specify it giving strings
+    containing file paths to the INCLUDE_LOCAL_FILES tuple:
```
Comment (sils1297) [inline coalib/bears/Bear.py]: always relative?

File: coalib/bears/Bear.py [hunk 11]
```diff
@@ -160,6 +177,22 @@ def include_local_files(cls):
         return ((cls.INCLUDE_LOCAL_FILES,) if isinstance(
             cls.INCLUDE_LOCAL_FILES, str) else cls.INCLUDE_LOCAL_FILES)
 
+    @classproperty
+    def can_detect(cls):
+        """
+        :return: A tuple with CAN_DETECTs.
+        """
+        return ((cls.CAN_DETECT,) if isinstance(
+            cls.CAN_DETECT, str) else cls.CAN_DETECT)
```
Comment (sils1297) [inline coalib/bears/Bear.py]: would you mind including `can_fix` implicitly? As check is a superset of fix.

Also I think we should rather use sets.

Comment (sils1297) [inline coalib/bears/Bear.py]: while we're at it: if we switch to sets we can drop the properties as theres no confusion about any lonely comma, you always use a set.

File: coalib/bears/Bear.py [hunk 12]
```diff
@@ -84,6 +113,13 @@ class Bear(Printer, LogPrinter):
 
     LANGUAGES = ()
     REQUIREMENTS = ()
+    AUTHOR = ""
+    AUTHOR_EMAIL = ""
+    MAINTAINERS = ()
+    MAINTAINERS_EMAILS = ()
+    PLATFORMS = 'any'
+    LICENSE = ""
+    INCLUDE_LOCAL_FILES = ()
```
Comment (sils1297) [inline coalib/bears/Bear.py]: can you use aset instead of tuples for all *new* attributes?

File: coalib/bears/Bear.py [hunk 13]
```diff
@@ -84,6 +99,13 @@ class Bear(Printer, LogPrinter):
 
     LANGUAGES = ()
     REQUIREMENTS = ()
+    AUTHOR = {}
+    AUTHOR_EMAIL = {}
+    MAINTAINERS = {}
+    MAINTAINERS_EMAILS = {}
+    PLATFORMS = {'any'}
+    LICENSE = ''
```
Comment (Adrianzatreanu) [inline coalib/bears/Bear.py]: @sils1297 i made all sets besides License. I think that theres no point this should be a set and not a string :)

File: coalib/bears/Bear.py [hunk 14]
```diff
@@ -84,6 +99,13 @@ class Bear(Printer, LogPrinter):
 
     LANGUAGES = ()
     REQUIREMENTS = ()
+    AUTHOR = {}
```
Comment (sils1297) [inline coalib/bears/Bear.py]: if it's a set you'll want to have AUTHORS rather, right?

File: coalib/bears/Bear.py (hunk 13 above)
Comment (AbdealiJK) [inline coalib/bears/Bear.py]: Why is LANGUAGES not a set ?

Comment (Adrianzatreanu) [inline coalib/bears/Bear.py]: it is not made by me, it was there before this PR. i plan on making it a set too, but different commit :)

File: coalib/bears/Bear.py [hunk 15]
```diff
@@ -119,8 +119,8 @@ def supported_languages(cls):
         """
         :return: The languages supported by the bear.
         """
-        return (cls.LANGUAGES if isinstance(
-            cls.LANGUAGES, tuple) else (cls.LANGUAGES,))
+        return ((cls.LANGUAGES,) if isinstance(
+            cls.LANGUAGES, str) else cls.LANGUAGES)
```
Comment (sils1297) [inline coalib/bears/Bear.py]: ahm no, you'll want to remove this function and make it a set :)

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2338/commits/5f1ad91c14ac6dc6bf7df58f125e1f39ed539964 https://github.com/coala-analyzer/coala/pull/2338/commits/413301e83012a10597097d74dbfe747a6d911548

**Comment by Adrianzatreanu** [comment]
reack a57d502

**Comment by Adrianzatreanu** [comment]
@rultor merge

**Pull request merged**